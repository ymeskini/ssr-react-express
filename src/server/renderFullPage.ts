type Params = {
  meta: string;
  body: string;
  style: string;
  preloadedState: string;
  nonce: string;
  scripts: string;
};

const escape = (str: string) => {
  return str
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
};

export const renderFullPage = ({ meta, body, style, scripts, preloadedState, nonce }: Params) => {
  return `<!DOCTYPE html>
    <html lang="fr">
      <head>
        <meta charSet="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="Description" content="PEX">
        ${meta}
        ${style}
      </head>
      <body>
        ${body}
        <script nonce="${nonce}" id="initial-data" type="text/plain" data-json="${escape(
    preloadedState
  )}"></script>
        ${scripts}
      </body>
    </html>
  `.trim();
};
