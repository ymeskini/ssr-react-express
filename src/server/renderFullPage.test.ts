import { renderFullPage } from './renderFullPage';

test('should create safe html', () => {
  const meta = '<meta>meta</meta>';
  const style = '<style>style</style>';
  const body = '<body>body</body>';
  const preloadedState = JSON.stringify({
    arr: ['foo', 'bar'],
    obj: {
      foo: 1,
      bar: 1
    },
    str: 'string'
  });
  const scripts = '<script>window</script>';
  const nonce = 'nonce';

  expect(renderFullPage({ meta, style, body, preloadedState, scripts, nonce })).toMatchSnapshot();
});
