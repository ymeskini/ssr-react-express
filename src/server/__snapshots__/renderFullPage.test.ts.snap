// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create safe html 1`] = `
"<!DOCTYPE html>
    <html lang=\\"fr\\">
      <head>
        <meta charSet=\\"UTF-8\\" />
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
        <meta name=\\"Description\\" content=\\"PEX\\">
        <meta>meta</meta>
        <style>style</style>
      </head>
      <body>
        <body>body</body>
        <script nonce=\\"nonce\\" id=\\"initial-data\\" type=\\"text/plain\\" data-json=\\"{&quot;arr&quot;:[&quot;foo&quot;,&quot;bar&quot;],&quot;obj&quot;:{&quot;foo&quot;:1,&quot;bar&quot;:1},&quot;str&quot;:&quot;string&quot;}\\"></script>
        <script>window</script>
      </body>
    </html>"
`;
